### PR TITLE
GitHub Actions CI workflow

### PR DESCRIPTION
**Pull request by timbussmann**
Setting up a github actions CI workflow that provides a valid MongoDB server to run the tests against.

While the windows image comes with MongoDB pre-installed, we require a server running with a replica set configuration, therefore some hackery is required to stop the existing server and restart it with the required settings.
For Linux workflows, we can use the "container services" to start MongoDB in a child container with the necessary replica set configuration.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        with:
+          fetch-depth: 0
+      - name: Checkout for Dependabot
+        if: github.event_name == 'pull_request_target'
+        uses: actions/checkout@v2.3.4
+        with:
+          ref: "refs/pull/${{ github.event.number }}/merge"
+          fetch-depth: 0
+      - name: Setup .NET SDK
+        uses: actions/setup-dotnet@v1.8.1
```
Comment (DavidBoike) [inline .github/workflows/ci.yml:46]: It updated today.
```suggestion
        uses: actions/setup-dotnet@v1.8.2
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        with:
+          dotnet-version: 5.0.x
+      - name: Setup .NET Core 3.1 runtime
+        uses: actions/setup-dotnet@v1.8.1
```
Comment (DavidBoike) [inline .github/workflows/ci.yml:50]:
```suggestion
        uses: actions/setup-dotnet@v1.8.2
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        uses: supercharge/mongodb-github-action@1.6.0
+        with:
+          mongodb-replica-set: tr0
+      - name: Run tests
+        uses: Particular/run-tests-action@v1.0.0
+      - name: Teardown infrastructure (Windows)
+        if: ${{ always() && matrix.name == 'Windows' }}
+        shell: pwsh
+        run: |
+          mongo --eval "db.getSiblingDB('admin').shutdownServer()"
```
Comment (DavidBoike) [inline .github/workflows/ci.yml:88]: Is this even needed? The agent is just going to be reset anyway, so wouldn't this just provide an opportunity for something to fail that doesn't point to a problem with our code? Or is it important to make sure the server can shut down?

**Review by DavidBoike — COMMENTED** [review]
Looks pretty good to me

File: .github/workflows/ci.yml (hunk 3 above)
Comment (timbussmann) [reply .github/workflows/ci.yml:88]: agreed, this doesn't seem to be needed as github hosted agents are fully reset between each use.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            name: Windows
+          - os: ubuntu-20.04
+            name: Linux
+      fail-fast: false
+    steps:
+      - name: Check for secrets
+        env:
+          SECRETS_AVAILABLE: ${{ secrets.SECRETS_AVAILABLE }}
+        shell: pwsh
+        run: exit $(If ($env:SECRETS_AVAILABLE -eq 'true') { 0 } Else { 1 })
```
Comment (soujay) [inline .github/workflows/ci.yml:33]:
```suggestion     
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        uses: actions/upload-artifact@v2.2.2
+        with:
+          name: NuGet packages
+          path: nugets/
+          retention-days: 7
+      - name: Stop built-in MongoDB server (Windows)
+        if: matrix.name == 'Windows'
+        shell: pwsh
+        run:  |
+          Stop-Service -Name "MongoDB"
```
Comment (soujay) [inline .github/workflows/ci.yml:66]:
```suggestion
          # Need to stop default MongoDB server as it does not support replication
          Stop-Service -Name "MongoDB"
```

**Comment by DavidBoike** [comment]
If secrets aren't necessary then the workflow can be simplified down to this template:

https://github.com/Particular/.github/blob/main/workflow-templates/ci.yml

(No big nasty if statement, single checkout)

**Review by helenktsai — APPROVED** [review]

**Pull request merged**